feat: AMOLED Dark Mode

abughalib opened the pull request:
### What
AMOLED Dark Mode

Changelog:
- `feat`: Added AMOLED Dark Mode.
- `feat`: Added Accent Color Picker
- `feat`: Added Text Contrast
- `fix`: Added AMOLED Test
- `fix`: SnackBar Background
### Screen Recording

https://user-images.githubusercontent.com/33085535/221379106-f0fe248c-0be4-4d25-a699-f2d44c58a658.mp4

Note:
Accent Color Picker is only available in AMOLED Dark Mode.

## Fixes
#3690

abughalib committed:
ignored some generated files

abughalib committed:
init amoled mode

abughalib committed:
color scheme

abughalib committed:
test updated for amoled

abughalib committed:
added accent color picker

abughalib committed:
localized accent color picker

abughalib committed:
added color provider

abughalib committed:
Dark Mode changed to Theme mode

abughalib committed:
Map key iteration only

AshAman999 commented on packages/smooth_app/lib/l10n/app_en.arb:
```diff
@@ -258,7 +258,7 @@
         "description": "The title of the Settings page"
     },
-    "darkmode": "Dark mode",
+    "darkmode": "Theme mode",
```
```suggestion
    "darkmode": "Theme",
```

would be more subtle ?

abughalib replied:
Makes sense, thanks.

AshAman999 commented on packages/smooth_app/lib/themes/color_provider.dart:
```diff
@@ -0,0 +1,15 @@
+import 'package:flutter/material.dart';
+import 'package:smooth_app/data_models/user_preferences.dart';
+
+class ColorProvider with ChangeNotifier {
```
Would love to see some comments in this file, would be helpful for other folks

abughalib replied:
You want comments in 10 lines of code, of which I think 100% of names are self-explanatory.
_Don't worry If I die anyone would be able to understand the code or rewrite it in 1 minutes_

abughalib replied:
Ok, I would add it.

AshAman999 reviewed (dismissed):
Had  a look and the changes looks good, , just few minor comments

Not sure but we can surely avoid commiting the auto generated plugin code,
And if you could fix the merge conflicts and undo the changes for the auto generated codes, we would be good to go then

abughalib committed:
Update packages/smooth_app/lib/l10n/app_en.arb

Co-authored-by: Aman Raj <57723319+AshAman999@users.noreply.github.com>

abughalib committed:
Merge branch 'develop' of https://github.com/abughalib/smooth-app into #3690

abughalib committed:
Merge branch '#3690' of https://github.com/abughalib/smooth-app into #3690

abughalib committed:
removed external dep

abughalib committed:
updated golden test

abughalib committed:
added comments

abughalib committed:
using shade fun

abughalib committed:
Merge branch 'develop' into #3690

M123-dev commented:
Heyy @abughalib as far as I looked it looks good codewise, but I am not sure about the color picker. We had one before release to test multiple colors, and we had a selection of colors to choose from for a while (blue, green and brown I think)

But when we got support from a professional UX designer we removed them in favour of a standard color to keep some branding guidelines, also the website now uses the same color scheme

Is there a way to intigrate the brownish color into the amoled scheme


@teolemon do we still have support from the designer

abughalib commented:
@M123-dev Yes, I can make the brownish color default or have that only.
That is what I did initially but after going through lots of Apps that supported AMOLED Theme, they were using one of those colors (Blue, Rust, Teal), and none of them was brown and that was the reason for me to add it.
All of the colors that are added to the color picker are from Windows Accent Colors.

If you want me to remove other colors and have the default only, I would do that.

adilwahla reviewed (approved):
looks amazing

M123-dev commented:
To keep in here what we talked about: Keep this additional color switch but we should make a brownish color the default one to not confuse anyone. We'll keep the branding unless you change it



Also, another idea which came to my mind. Maybe we could name this color setting something with high contrast and make it into an accessibility function. But we can think about that later in more detail

abughalib commented:
@M123-dev 
I'm currently using white hex code: 0xFFE1E1E1, which does qualify for High Contrast.

After doing some research on the Internet and asking almost the same repeated questing from Bing AI, I figured out the contrast ratio for a good viewing experience is supported to be greater than **4.5:1:1** and relative luminance between **0.2** and **0.8**, Currently, my implementation has **16.05:1:1** and the relative luminance is **0.75**.

For High contrast, these numbers should be more than __7:1:1__ up to __21:1:1__.

**Should I add something to increase the contrast from 16 to 21?**

_All of these research are based on the pure back (#000000) background_

abughalib commented:
I'm using these tools for calculations.
[WCAG Color Contrast (App)](https://developer.paciellogroup.com/color-contrast-checker)
[Contrast Ratio by LeaVerou (Webpage)](https://contrast-ratio.com)

M123-dev commented:
I'd say lets use the brownish color as a normal one, without focus on high contrast. Of course, it should be visible though

And we can add one or two color specifically for high contrast

(Feel free to  do that in a seperate PR, we already have some conflicts here :D)

abughalib committed:
color scheme localization

abughalib committed:
color scheme localization

abughalib committed:
added textContrast

abughalib committed:
Merge branch 'develop' of https://github.com/abughalib/smooth-app into #3690

abughalib committed:
color contrast app settings

abughalib committed:
Merge branch '#3690' of https://github.com/abughalib/smooth-app into #3690

abughalib committed:
merged upstream

abughalib committed:
typo fix

abughalib committed:
Merge branch 'develop' into #3690

M123-dev commented:
@abughalib what did you change since my last comment

abughalib commented:
@M123-dev 
The Light Brown Color is the default now.

I added Text Contrast, you can check out the new Video on top and added the text style to List Items (the ones used on History and on product search Result), and DropDownMenuItem in App Settings so that they would all use the same text contrast level.

M123-dev commented:
Could you add screenshots of the default color in amoled from a few different pages ;-D

abughalib commented:
Sure, Give me a few minutes.

abughalib commented:
@M123-dev 
I tried getting to most of the pages.
https://user-images.githubusercontent.com/33085535/222203417-f121a823-e3ca-4cf4-8a1d-b1e52e539db5.mp4

I have noticed, the snack bar message has a white background that makes the text invisible.
I would fix that.

abughalib committed:
Merge branch 'develop' into #3690

abughalib committed:
Merge branch '#3690' of https://github.com/abughalib/smooth-app into #3690

abughalib committed:
Merge branch 'openfoodfacts:develop' into #3690

abughalib committed:
Merge branch '#3690' of https://github.com/abughalib/smooth-app into #3690

M123-dev reviewed (commented):
Heyy @abughalib thanks again for all this effort, I cant wait to merge it. Just one last minor comment regarding the snack bars.

You hardcoded the color values in every snackbar, which is okay, but will likely break the next snackbar which we'll add. Therefor could you change the global `snackBarTheme` in `smooth_theme.dart` to reflect your color changes

abughalib commented:
RE: @M123-dev 
My original thought was to create a SmoothSnackBar (widget for all SnackBar).
But in order to finish the PR quickly I just added the background and text color to every Snackbar.
I can still do what you suggested or create a new Widget like I initially planned.
Let me know whichever you like and I would start working on it.

M123-dev commented:
I don't think a own widget is needed here @abughalib , I think when you change the color (snackBarTheme) in our SmoothTheme (lib/themes/smooth_theme.dart) all snackbars should use this automatically

abughalib commented:
I would do that.

M123-dev commented:
Perfect

abughalib commented:
One more thing crossed my mind, I'm thinking of adding color as well to the DropDownMenuItem.
It could be something like...
![image](https://user-images.githubusercontent.com/33085535/223788678-756cadab-50dc-4590-8abb-85abd621fd7c.png) Magenta
![image](https://user-images.githubusercontent.com/33085535/223788935-e97f5ee5-370e-4435-9f19-a66fb955d92e.png) Teal

So on...

Let me know if you think it any good.

M123-dev commented:
Sounds good to me if it doesn't take too long this PR is already open for a long time

abughalib commented:
I would try to finish it quickly and efficiently.

abughalib committed:
Merge branch 'openfoodfacts:develop' into #3690

abughalib committed:
Merge branch 'develop' of https://github.com/abughalib/smooth-app into #3690

abughalib committed:
Merge branch '#3690' of https://github.com/abughalib/smooth-app into #3690

abughalib committed:
added color to DropDownMenuItems

abughalib committed:
SnackBar text, contrast level

abughalib committed:
text contrast change to pass WCAG2.1

abughalib committed:
snackBar Color Change

abughalib commented:
## Changes from Last Commit.

- Removed the hardcoded color from the snack bar.
- Added background to the snack bar.
- Added Accent Color Preview.
- Added Text Contrast Preview.

## Screenshots
<details>
Accent Color Preview

![Screenshot_1678373199](https://user-images.githubusercontent.com/33085535/224069525-557d2297-080f-49dc-984e-64069697b0f1.png)

Text Contrast Preview
![Screenshot_1678373194](https://user-images.githubusercontent.com/33085535/224069446-6c2a98d9-10e6-4f67-a2a5-62f604930959.png)

Video of SnackBar & Others.

[snackbar.webm](https://user-images.githubusercontent.com/33085535/224074513-aa6bdeb8-13b4-447c-a526-da98796247d7.webm)

</details>

M123-dev reviewed (approved):
👍🏼
